src/replication.c

```diff
@@ -1972,12 +1993,26 @@ void readSyncBulkPayload(connection *conn) {
 
         /* RDB loading succeeded if we reach this point. */
         if (server.repl_diskless_load == REPL_DISKLESS_LOAD_SWAPDB) {
-            /* Delete the backup databases we created before starting to load
-             * the new RDB. Now the RDB was loaded with success so the old
-             * data is useless. */
-            disklessLoadDiscardBackup(diskless_load_backup, empty_db_flags);
+            /* We will soon swap main db with tempDb and replicas will start
+             * to apply data from new master, we must discard the cached
+             * master structure and force resync of sub-replicas. */
+            replicationAttachToNewMaster();
+
+            serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Swapping active DB with loaded DB");
+            swapMainDbWithTempDb(diskless_load_tempDb);
+
+            moduleFireServerEvent(REDISMODULE_EVENT_REPL_ASYNC_LOAD,
+                        REDISMODULE_SUBEVENT_REPL_ASYNC_LOAD_COMPLETED,
+                        NULL);
+
+            /* Delete the old db as it's useless now. */
+            disklessLoadDiscardTempDb(diskless_load_tempDb);
+            serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Discarding old DB in background");
         }
 
+        /* Inform about db change, as replication was diskless and didn't cause a save. */
+        server.dirty++;
+
         /* Verify the end mark is correct. */
         if (usemark) {
             if (!rioRead(&rdb,buf,CONFIG_RUN_ID_SIZE) ||
```

Comment:
@MeirShpilraien noticed that the failure here can happen after we already announced ASYNC_LOAD_COMPLETED, swapped the databases and discarded the backup.
@eduardobr can you please look into it and issue a fix?